File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>>ans = new ArrayList<>();
    public void helper(int[] candidates, int target, int index, List<Integer> lst)
    {
        if(target == 0){
            ans.add(new ArrayList<>(lst));
            return;
        }
        if(index == candidates.length || target < 0)
            return;
        
        helper(candidates, target, index+1, lst);
        lst.add(candidates[index]);
        helper(candidates, target - candidates[index], index, lst);
        lst.remove(lst.size() - 1);
            
            
            
    }
    

    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        helper(candidates, target, 0, new ArrayList<>() );
        return ans;
        
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def __init__(self):
        self.ans = []

    def helper(self, candidates, target, index, lst):
        if target == 0:
            self.ans.append(lst[:])
            return
        if index == len(candidates) or target < 0:
            return
        
        # Exclude the current candidate
        self.helper(candidates, target, index + 1, lst)
        
        # Include the current candidate
        lst.append(candidates[index])
        self.helper(candidates, target - candidates[index], index, lst)
        lst.pop()
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        self.ans = []  # Clear the ans list before starting a new computation
        self.helper(candidates, target, 0, [])
        return self.ans